utils/fixed_decimal/src/decimal.rs

```diff
@@ -1054,6 +1068,113 @@ impl FixedDecimal {
         Self::from_str(formatted)
     }
 
+    /// Rounds the number according to the power (`n`) of 10 position.
+    ///
+    /// NOTE: This may end up adding a digit to the left!
+    ///
+    /// NOTE: This will not change the number of significant digits, it simply exists
+    /// to *round* them (and will typically reduce the size of `self.digits`)
+    ///
+    /// This function is responsible for fixing `digits`, `magnitude`, and `upper_magnitude`.
+    /// It will only modify upper_magnitude when it is not large enough to fit the rounded number.
+    /// The caller may fix up `lower_magnitude` by whatever scheme it desires
+    pub fn round(&mut self, n: u16, mode: RoundingMode) -> Result<(), Error> {
+        /// For incrementing the FixedDecimal absolutely by 1.
+        fn increment_abs_by_one(dec: &mut FixedDecimal) {
+            if dec.magnitude < 0 {
+                return;
+            }
+
+            for i in 0..dec.digits.len() {
+                dec.digits[i] += 1;
+                if dec.digits[i] >= 10 {
+                    dec.digits[i] = 0;
+                    continue;
+                } else {
+                    return;
+                }
+            }
+
+            dec.digits.push(1);
+            dec.upper_magnitude += 1;
```

Comment:
Issue: This function leaves the `FixedDecimal` in a state where not all invariants are satisfied. Please add

```
        #[cfg(debug_assertions)]
        decimal.check_invariants();
```

to the bottom of this function so you can see errors that you can fix. (You need to shift the trailing zeros out of `dec.digits`.)